### PR TITLE
etcdserver: check -initial-cluster in join case

### LINKED ISSUE #2396: Error: "invalid memory address or nil pointer dereference" when trying to add new node to existing cluster
Hi,

I'm currently experiencing an issue when trying to add a new node to an existing cluster. I'm running 2.0.3 from within a Docker container. My log looks somewhat like the following:

Feb 27 14:52:58 ip-10-21-1-94.ec2.internal systemd[1]: Starting etcd container...
Feb 27 14:52:58 ip-10-21-1-94.ec2.internal writer_process[4938]: etcd
Feb 27 14:52:58 ip-10-21-1-94.ec2.internal writer_process[4945]: etcd
Feb 27 14:52:58 ip-10-21-1-94.ec2.internal writer_process[4953]: Pulling repository quay.io/coreos/etcd
Feb 27 14:52:59 ip-10-21-1-94.ec2.internal writer_process[4953]: Status: Image is up to date for quay.io/coreos/etcd:v2.0.3
Feb 27 14:52:59 ip-10-21-1-94.ec2.internal systemd[1]: Started etcd container.
Feb 27 14:52:59 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:52:59 etcd: listening for peers on http://0.0.0.0:2380
Feb 27 14:52:59 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:52:59 etcd: listening for peers on http://0.0.0.0:7001
Feb 27 14:52:59 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:52:59 etcd: listening for client requests on http://0.0.0.0:2379
Feb 27 14:52:59 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:52:59 etcd: listening for client requests on http://0.0.0.0:4001
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:53:00 etcdserver: could not get cluster response from http://10.21.1.43:2380: Get http://10.21.1.43:2380/members: net/http: request canceled while waiting for connection
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:53:00 etcdserver: name = ip-10-21-1-94.ec2.internal
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:53:00 etcdserver: data dir = /opt/etcd
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:53:00 etcdserver: member dir = /opt/etcd/member
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:53:00 etcdserver: heartbeat = 100ms
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:53:00 etcdserver: election = 1000ms
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:53:00 etcdserver: snapshot count = 10000
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:53:00 etcdserver: advertise client URLs = http://10.21.1.94:2379,http://10.21.1.94:4001
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: panic: runtime error: invalid memory address or nil pointer dereference
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: [signal 0xb code=0x1 addr=0x0 pc=0x4a5b5d]
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: goroutine 16 [running]:
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: runtime.panic(0x7dc1e0, 0xa63493)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/runtime/panic.c:279 +0xf5
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: github.com/coreos/etcd/etcdserver.startNode(0xc20803a580, 0x0, 0x0, 0x0, 0xc208044f80, 0x0, 0x0, 0x10, 0x10)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /Users/unihorn/gopath/src/github.com/coreos/etcd/scripts/release/etcd/gopath/src/github.com/coreos/etcd/etcdserver/raft.go:78 +0xdd
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: github.com/coreos/etcd/etcdserver.NewServer(0xc20803a580, 0xc20803a580, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /Users/unihorn/gopath/src/github.com/coreos/etcd/scripts/release/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:171 +0x6bc
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: github.com/coreos/etcd/etcdmain.startEtcd(0xc20807c000, 0x0, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /Users/unihorn/gopath/src/github.com/coreos/etcd/scripts/release/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:162 +0x1013
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: github.com/coreos/etcd/etcdmain.Main()
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /Users/unihorn/gopath/src/github.com/coreos/etcd/scripts/release/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:58 +0x1c5
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: main.main()
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /Users/unihorn/gopath/src/github.com/coreos/etcd/scripts/release/etcd/gopath/src/github.com/coreos/etcd/main.go:48 +0x1ab
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: goroutine 19 [finalizer wait]:
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: runtime.park(0x4131d0, 0xa67fd8, 0xa66209)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/runtime/proc.c:1369 +0x89
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: runtime.parkunlock(0xa67fd8, 0xa66209)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/runtime/proc.c:1385 +0x3b
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: runfinq()
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/runtime/mgc0.c:2644 +0xcf
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: runtime.goexit()
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/runtime/proc.c:1445
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: goroutine 20 [syscall]:
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: os/signal.loop()
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/os/signal/signal_unix.go:21 +0x1e
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: created by os/signal.init·1
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/os/signal/signal_unix.go:27 +0x32
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: goroutine 28 [IO wait]:
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net.runtime_pollWait(0x7fbec1e58938, 0x72, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/runtime/netpoll.goc:146 +0x66
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net.(*pollDesc).Wait(0xc20807efb0, 0x72, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/fd_poll_runtime.go:84 +0x46
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net.(*pollDesc).WaitRead(0xc20807efb0, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/fd_poll_runtime.go:89 +0x42
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net.(*netFD).Read(0xc20807ef50, 0xc208074000, 0x1000, 0x1000, 0x0, 0x7fbec1e573f0, 0xb)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/fd_unix.go:242 +0x34c
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net.(*conn).Read(0xc2080380a0, 0xc208074000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/net.go:122 +0xe7
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: github.com/coreos/etcd/pkg/transport.timeoutConn.Read(0x7fbec1e58f30, 0xc2080380a0, 0x12a05f200, 0x12a05f200, 0xc208074000, 0x1000, 0x1000, 0x7fbec1e590b8, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /Users/unihorn/gopath/src/github.com/coreos/etcd/scripts/release/etcd/gopath/src/github.com/coreos/etcd/pkg/transport/timeout_conn.go:43 +0x16f
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: github.com/coreos/etcd/pkg/transport.(*timeoutConn).Read(0xc2080405e0, 0xc208074000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: <autogenerated>:21 +0xcc
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net/http.noteEOFReader.Read(0x7fbec1e59068, 0xc2080405e0, 0xc20803a738, 0xc208074000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:1203 +0x72
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net/http.(*noteEOFReader).Read(0xc208040600, 0xc208074000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: <autogenerated>:124 +0xca
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: bufio.(*Reader).fill(0xc2080049c0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/bufio/bufio.go:97 +0x1b3
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: bufio.(*Reader).Peek(0xc2080049c0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/bufio/bufio.go:132 +0x101
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net/http.(*persistConn).readLoop(0xc20803a6e0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:782 +0x95
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: created by net/http.(*Transport).dialConn
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:600 +0x93f
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: goroutine 26 [runnable]:
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net/http.(*persistConn).readLoop(0xc20803a2c0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:868 +0x829
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: created by net/http.(*Transport).dialConn
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:600 +0x93f
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: goroutine 27 [select]:
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net/http.(*persistConn).writeLoop(0xc20803a2c0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:885 +0x38f
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: created by net/http.(*Transport).dialConn
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:601 +0x957
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: goroutine 29 [select]:
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net/http.(*persistConn).writeLoop(0xc20803a6e0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:885 +0x38f
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: created by net/http.(*Transport).dialConn
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:601 +0x957
Feb 27 14:53:01 ip-10-21-1-94.ec2.internal systemd[1]: etcd.service: main process exited, code=exited, status=2/INVALIDARGUMENT
Feb 27 14:53:01 ip-10-21-1-94.ec2.internal writer_process[5026]: etcd
Feb 27 14:53:01 ip-10-21-1-94.ec2.internal systemd[1]: Unit etcd.service entered failed state.

The existing cluster consists of 2 nodes, my etcd.service looks like this:

[Service]
EnvironmentFile=/etc/environment
TimeoutStartSec=0
SyslogIdentifier=writer_process
ExecStartPre=-/usr/bin/docker kill etcd
ExecStartPre=-/usr/bin/docker rm etcd
ExecStartPre=/usr/bin/docker pull quay.io/coreos/etcd:v2.0.3
ExecStart=/bin/bash -c "/usr/bin/docker run \
  -p 2379:2379 \
  -p 2380:2380 \
  -p 4001:4001 \
  -p 7001:7001 \
  --name etcd \
  -v /opt/etcd:/opt/etcd \
  -v /usr/share/ca-certificates/:/etc/ssl/certs \
  quay.io/coreos/etcd:v2.0.3 \
  -data-dir /opt/etcd \
  -name %H \
  -listen-client-urls http://0.0.0.0:2379,http://0.0.0.0:4001 \
  -advertise-client-urls http://10.21.1.94:2379,http://10.21.1.94:4001 \
  -listen-peer-urls http://0.0.0.0:2380,http://0.0.0.0:7001 \
  -initial-advertise-peer-urls http://10.21.1.94:2380,http://10.21.1.94:7001 \
  -initial-cluster-state existing \
  -initial-cluster ip-10-21-1-43.ec2.internal=http://10.21.1.43:2380,ip-10-21-1-43.ec2.internal=http://10.21.1.43:7001,ip-10-21-2-166.ec2.internal=http://10.21.2.166:2380,ip-10-21-2-166.ec2.internal=http://10.21.2.166:7001"
ExecStop=/usr/bin/docker kill etcd

I thought at first it might be related to the "could not get cluster response", but a quick check from .94 works:

core@ip-10-21-1-94 /etc/systemd/system $ curl -L http://10.21.1.43:2380/members
[{"id":12394471189341718283,"peerURLs":["http://10.21.2.166:2380","http://10.21.2.166:7001"],"name":"ip-10-21-2-166.ec2.internal","clientURLs":["http://10.21.2.166:2379","http://10.21.2.166:4001"]},{"id":13661735980988760828,"peerURLs":["http://10.21.1.43:2380","http://10.21.1.43:7001"],"name":"ip-10-21-1-43.ec2.internal","clientURLs":["http://10.21.1.43:2379","http://10.21.1.43:4001"]}]

Any help would be hugely appreciated.

### PR DESCRIPTION
**Pull request by yichengq**
fixes #2396 

/cc @wsimmonds I have reproduced the panic and fixed it through this PR. Could you double-check it?

### CODE REVIEW

File: etcdserver/config.go [hunk 1]
```diff
@@ -58,10 +58,6 @@ func (c *ServerConfig) VerifyBootstrapConfig() error {
 		return fmt.Errorf("cannot use %x as member id", raft.None)
 	}
 
-	if c.DiscoveryURL == "" && !c.NewCluster {
-		return fmt.Errorf("initial cluster state unset and no wal or discovery URL found")
-	}
-
```
Comment (yichengq) [inline etcdserver/config.go]: The description of the error message is not accurate. And i think what it wants to check has been covered by `// Advertised peer URLs must match those in the cluster peer list` ten lines below.

**Comment by yichengq** [comment]
integration tests fixed. PTAL

**Comment by yichengq** [comment]
split existing and bootstrap verify. PTAL

File: etcdserver/config.go [hunk 2]
```diff
@@ -49,6 +49,17 @@ type ServerConfig struct {
...
+}
+
+func (c *ServerConfig) VerifyExistingConfig() error {
+	if c.DiscoveryURL != "" {
+		return fmt.Errorf(`initial cluster state "existing" but discovery URL found`)
+	}
+	return c.verifyConfig()
+}
+
+func (c *ServerConfig) verifyConfig() error {
```
Comment (xiang90) [inline etcdserver/config.go]: add comments to this one, what does it verify?

**Comment by yichengq** [comment]
PTAL

File: etcdserver/config_test.go [hunk 3]
```diff
@@ -29,74 +29,121 @@ func mustNewURLs(t *testing.T, urls []string) []url.URL {
...
+	}
+}
+
+func TestConfigVerifyExistingWithDiscoveryURLFail(t *testing.T) {
+	c := &ServerConfig{
+		NewCluster:   false,
+		DiscoveryURL: "http://127.0.0.1:4001/abcdefg",
+	}
+	if err := c.VerifyJoinExisting(); err == nil {
+		t.Errorf("got no error when one is expected")
```
Comment (xiang90) [inline etcdserver/config_test.go]: err = nil, want not nil

**Comment by xiang90** [comment]
LGTM.

File: etcdserver/config.go [hunk 4]
```diff
@@ -48,7 +48,41 @@ type ServerConfig struct {
...
+	}
+	if err := c.Cluster.Validate(); err != nil {
+		return err
+	}
+	hasMultiMembers := len(c.Cluster.MemberIDs()) > 1
+	// If initial cluster is unset, it only has the local member. So if
+	// it has multiple members, the initial cluster must be set.
+	if hasMultiMembers && c.DiscoveryURL != "" {
+		return fmt.Errorf("both initial cluster and discovery URL set")
+	}
```
Comment (yichengq) [inline etcdserver/config.go]: Make some change on the check here to make it fit current workload.

**Comment by yichengq** [comment]
PTAL

File: etcdserver/config.go [hunk 5]
```diff
@@ -48,7 +48,41 @@ type ServerConfig struct {
...
 // for things that should never happen.
-func (c *ServerConfig) VerifyBootstrapConfig() error {
+func (c *ServerConfig) VerifyBootstrap() error {
+	if err := c.verifyLocalMember(); err != nil {
+		return err
+	}
+	if err := c.Cluster.Validate(); err != nil {
+		return err
+	}
+	hasMultiMembers := len(c.Cluster.MemberIDs()) > 1
```
Comment (xiang90) [inline etcdserver/config.go]: this is too confusing. i guess we should simply remove this checking.

File: etcdserver/config.go [hunk 6]
```diff
@@ -48,7 +48,41 @@ type ServerConfig struct {
...
+	if err := c.Cluster.Validate(); err != nil {
+		return err
+	}
+	if c.DiscoveryURL != "" {
+		return fmt.Errorf(`initial cluster state "existing" but discovery URL found`)
+	}
+	return nil
+}
+
+// verifyLocalMember verifies that the member is valid and is listed
```
Comment (xiang90) [inline etcdserver/config.go]: the local member

File: etcdserver/config.go (hunk 5 above)
Comment (yichengq) [inline etcdserver/config.go]: k

**Comment by yichengq** [comment]
PTAL

File: etcdserver/config.go [hunk 7]
```diff
@@ -46,9 +46,39 @@ type ServerConfig struct {
...
+// case and returns an error for things that should never happen.
+func (c *ServerConfig) VerifyJoinExisting() error {
+	if err := c.verifyLocalMember(); err != nil {
+		return err
+	}
+	if err := c.Cluster.Validate(); err != nil {
+		return err
+	}
+	if c.DiscoveryURL != "" {
+		return fmt.Errorf(`initial cluster state "existing" but discovery URL found`)
```
Comment (xiang90) [inline etcdserver/config.go]: discovery URL should not be set when joining existing initial cluster.

**Pull request merged**